server/src/main/java/org/apache/cassandra/sidecar/livemigration/LiveMigrationFileDownloader.java

```diff
@@ -0,0 +1,688 @@
+                filesDeleted.incrementAndGet();
+            }
+        }
+
+        LOGGER.info("{} Deleted {} unwanted files from local.", logPrefix, filesDeleted.get());
+        LOGGER.info("{} Deleted {} unwanted directories from local.", logPrefix, directoriesDeleted.get());
+    }
+
+    /**
+     * Shortlists list of files and directories to download from source.
+     *
+     * @param instanceFilesListResponse list of files response received from source.
+     * @param successThreshold          success threshold to determine whether downloader should continue to download or not.
+     * @return List of instance files to download. Returned list will be empty if successThreshold meets or no files remained to download.
+     */
+    @VisibleForTesting
+    Future<List<InstanceFileInfo>> shortlistDownloadFiles(InstanceFilesListResponse instanceFilesListResponse,
+                                                          double successThreshold)
+    {
+        List<InstanceFileInfo> filesToDownload =
+        instanceFilesListResponse.getFiles().stream()
+                                 .filter(instanceFileInfo -> !Files.exists(localPath(instanceFileInfo.fileUrl, instanceMetadata)))
+                                 .collect(Collectors.toList());
+
+        // filter out directories while calculating download size
+        long downloadSize = filesToDownload.stream()
+                                           .filter(instanceFileInfo -> instanceFileInfo.fileType.equals(FileType.FILE))
+                                           .map(instanceFileInfo -> instanceFileInfo.size)
+                                           .reduce(0L, Long::sum);
+
+        if (downloadSize == 0)
+        {
+            // downloadSize of zero means that there are no files to download OR
+            // only empty directories and/or empty files left to create.
+            return Future.succeededFuture(filesToDownload);
+        }
+
+        long totalSize = instanceFilesListResponse.getTotalSize();
+
+        // Calculate percentage of data required to download
+        double toDownloadPercentage = ((totalSize - downloadSize) * 1.0) / totalSize;
+        LOGGER.info("{} Remaining size to download={} bytes", logPrefix, downloadSize);
+        LOGGER.info("{} Download percentage={}, success threshold={}", logPrefix, toDownloadPercentage, successThreshold);
+
+        if (toDownloadPercentage <= successThreshold)
+        {
+            return Future.succeededFuture(filesToDownload);
+        }
+
+        LOGGER.info("{} Download size met the successThreshold({}).{}",
+                    logPrefix,
+                    successThreshold,
+                    " Skipping download of " + filesToDownload.size() + " files of size " + downloadSize);
+        return Future.succeededFuture(Collections.emptyList());
+    }
+
+    @VisibleForTesting
+    Future<OperationStatus> sortBySizeAndDownload(List<InstanceFileInfo> instanceFileInfoList)
```

Comment:
Can you comment the reason of downloading the large files first? 
And.. my wondering.. if dictionaries are sorted the last, will the directories already be created on the downloading host when downloading the files?